Schedule concurrent tasks

Sagi-levi opened the pull request:
## Overview
This PR introduces a basic concurrent task scheduler implementation in Go, establishing the foundational structure and interfaces for task scheduling functionality.

### Scheduler abilities
The scheduler allows you to configure the maximum number of tasks that can run simultaneously, as well as the maximum number of tasks that can wait in the execution queue. Additionally, the scheduler stores task statistics and results atomically, preventing race conditions and data races.

The scheduler provides the ability to modify tasks properties, including editing task names, adjusting retry counts, and controlling logging. 
The main target is to provide a simple interface for end users to implement when interacting with the scheduler, an example can be examined at the test file.

### Development and arch guidelines 
Fan in, fan out pattern, taking care of the reading the same channel using multiple go routines.
Decorator providing the name, retry, logging dynamic adding by placing them inside the register function.

POW:  https://github.com/Sagi-levi/Task-Scheduler/pull/1#issuecomment-3096536145

Sagi-levi commented:
### Proof Of Work:
<details><summary>main.go</summary>
<p>

```go
func main() {
	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))
	scheduler, err := taskscheduling.New(40, 4)
	if err != nil {
		log.Fatal(err)
	}
	task := func() error {
		return nil
	}
	failingTask := func() error {
		return errors.New("error")
	}
	err = scheduler.Register(task, taskscheduling.WithRetry(4), taskscheduling.WithName("number1"), taskscheduling.WithLogging(logger))
	if err != nil {
		log.Fatal(err)
	}
	err = scheduler.Register(task, taskscheduling.WithRetry(2), taskscheduling.WithName("number2"))
	if err != nil {
		log.Fatal(err)
	}
	err = scheduler.Register(task, taskscheduling.WithLogging(logger))
	if err != nil {
		log.Fatal(err)
	}
	err = scheduler.Register(failingTask, taskscheduling.WithRetry(4), taskscheduling.WithName("number4"), taskscheduling.WithLogging(logger))
	if err != nil {
		log.Fatal(err)
	}
	scheduler.Run()
	scheduler.Stop()
	scheduler.Summary()
}

``` 

</p>
</details> 
In the codebook above, we can see the implementation of the package functionality by a potential user:

Configurable scheduler 
Task creating
Task registering to the scheduler
high-order functions enabling retry, task naming, and logging enabling 

Running results:
<details><summary>logs</summary>
<p>

```
time=2025-07-31T14:36:14.826+03:00 level=INFO msg=registered
time=2025-07-31T14:36:14.827+03:00 level=INFO msg=registered
time=2025-07-31T14:36:14.827+03:00 level=INFO msg=registered
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number1 attempt 1"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number1 attempt 1"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number1 attempt 2"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number1 attempt 2"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number1 attempt 3"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number1 attempt 3"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number1 attempt 4"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number1 attempt 4"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number4 attempt 1"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="failed task-number4 attempt 1"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number4 attempt 1"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number4 attempt 2"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="failed task-number4 attempt 2"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number4 attempt 2"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number4 attempt 3"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="failed task-number4 attempt 3"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number4 attempt 3"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-number4 attempt 4"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="failed task-number4 attempt 4"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-number4 attempt 4"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="running task-task-c2a5c116-c91c-47d6-98fd-af77f8565673 attempt 1"
time=2025-07-31T14:36:14.827+03:00 level=INFO msg="finished task-task-c2a5c116-c91c-47d6-98fd-af77f8565673 attempt 1"
╔═════════════════════════════════════════════════════════╗
║               Scheduler Execution Report                ║
╠═════════════════════════════════════════════════════════╣
║ Registered tasks        :     4                          ║
║ Total executions        :    11                          ║
║ Successful executions   :     7                          ║
║ Failed executions       :     4                          ║
╠═════════════════════════════════════════════════════════╣
║                    Per-Task Results                     ║
╠═════════════════════════════════════════════════════════╣
║ number1                                   rep:  0 ➜ OK   ║
║ number1                                   rep:  1 ➜ OK   ║
║ number2                                   rep:  0 ➜ OK   ║
║ number2                                   rep:  1 ➜ OK   ║
║ number1                                   rep:  2 ➜ OK   ║
║ number1                                   rep:  3 ➜ OK   ║
║ number4                                   rep:  0 ➜ ERR  ║
║ number4                                   rep:  1 ➜ ERR  ║
║ number4                                   rep:  2 ➜ ERR  ║
║ number4                                   rep:  3 ➜ ERR  ║
║ task-c2a5c116-c91c-47d6-98fd-af77f8565673 rep:  0 ➜ OK   ║
╚═════════════════════════════════════════════════════════╝
``` 

</p>
</details> 

From the logs above, we can assume several things:
Tasks were executed simultaneously and not in chronological order.
The high-order functions control the outcome of the logging execution, the task name and retries.
We can observe the results with the summary function shows us what happened during our run.

arieltod commented:
hi @Sagi-levi 🥳

Great job for your first PR in OneLayer 👍
![gif](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExZjdiNDd5djQ5M2VycXN1eGowa2VnazF3dHJ0Z3cya3Zqdzg1MWY1cyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/L0O3TQpp0WnSXmxV8p/giphy.gif)

Before reviewing the actual code, I first want to verify that this PR tells the story of the exercise you were given. 
So my first comments here will be mainly about the supporting elements of the PR- title, description, commits, code structure and design, used patterns, proof of work etc.

We’ll discuss this in our meeting tomorrow 🤞.

Thanks!

Sagi-levi committed:
go.mod: init task-scheduler

This commit inits the task-scheduler. This commit also creates a gitignore file
that ignores files and dirs created while developing and doesn't affect the
code functionality or created by the IDE.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -0,0 +1,212 @@
+	"sync/atomic"
+)
+
+// DefaultTaskChannelBuffer defines the default size of the task channel buffer used in TaskSchedule initialization.
```
Looks like a typo in const name, the const is not exported anymore. Please rename in the comment.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -0,0 +1,212 @@
+// 20 Was chosen to allow a minimal but reasonable buffer
+const defaultTaskChannelBuffer = 20
+
+// DefaultConcurrentWorkerLimit defines the default maximum number of concurrent workers allowed in a TaskSchedule.
```
Looks like a typo in const name, the const is not exported anymore. Please rename in the comment.

arieltod commented on scheduler/higherOrder.go:
```diff
@@ -0,0 +1,22 @@
+}
+func WithName(name string) HighOrderTaskHandler {
+	return func(task *taskHandler) {
+		task.Name = name
```
hi @Sagi-levi, this line of code throws an error in IDE, please fix it :) 
```shell
# scheduler/scheduler
scheduler/higherOrder.go:14:8: task.Name undefined (type *taskHandler has no field or method Name, but does have field name)
```

arieltod commented on scheduler/tasker.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func (t *taskHandler) runTask(ctx context.Context) error {
+	return t.fn.Run(ctx)
```
@Sagi-levi reconsider the naming of the interface methods. 
For example the usage of this func in the code looks like:
```go
task.runTask(context.Background())
``` 
Improving the naming could enhance clarity and readability.

Sagi-levi replied:
Done, look like this now. @arieltod 
```go
func (t *taskHandler) run(ctx context.Context) error {
	return t.fn.Run(ctx)
}


task.run(context.Background())
```

arieltod commented on scheduler/tasker.go:
```diff
@@ -0,0 +1,38 @@
+	return t.fn.Run(ctx)
+}
+
+type taskRes struct {
```
Please consider better naming to your struct types, in this case `res` instead of `taskRes`.

Sagi-levi replied:
Changed to :
```go
type res struct {

```
 res can be too generic in my opinion, in case of "open to changes" principle we would someday have another result, and we will need to specify our result struct.

arieltod commented on main.go:
```diff
@@ -0,0 +1,29 @@
+	exampleTask2 := &scheduler.ExampleTask{}
+	exampleTask3 := &scheduler.ExampleTask{}
+	exampleTask4 := &scheduler.ExampleTask{}
+	exampleTask5 := &scheduler.ExampleTaskFail{}
```
hi @Sagi-levi, 
We're usually not using the `main.go` file to show examples of packages, let's use `ExampleTest` instead. 
Check this doc: https://go.dev/blog/examples.

Sagi-levi replied:
Changed to 
```go
	Task1 := &scheduler.MyTask{}
	Task4 := &scheduler.MyTask{}
	Task3 := &scheduler.MyTask{}
	Task2 := &scheduler.MyTask{}
	Task5 := &scheduler.MyTask{Fail: true}
``` 
An example was added to the test file.

arieltod commented on scheduler/doc.go:
```diff
@@ -0,0 +1,45 @@
+/*
+Package scheduler provides a concurrent task execution framework for Go applications.
```
Hi @Sagi-levi, 
please review your package documentation using the `pkgsite` [tool](https://pkg.go.dev/about),
specifically this file contains some extra indentation.

Sagi-levi replied:
Done successfully.

arieltod replied:
Let's review this again with the `pkgsite` tool issue you had before.
Also please verify that the doc.go content is atomic along the PR commits too. 

For example in this specific commit the doc describes functionality that will be implemented in future commits. 

> retry mechanisms, logging capabilities, and worker pool management.

arieltod commented on scheduler/exampleTask.go:
```diff
@@ -0,0 +1,20 @@
+	"errors"
+)
+
+type ExampleTask struct {
```
hi @sagi,
consider implementing `Task` with interface, for this case with `ExampleTaskFail` it's a classic case of go Interface. 
Also- feel free to rename the struct `ExampleTask` to some prettier name like `Task` 👼

Sagi-levi replied:
Changed to implementing the task interface and the name changed to MyTask.

arieltod commented on scheduler/higherOrder.go:
```diff
@@ -0,0 +1,22 @@
+
+import "log/slog"
+
+type HighOrderTaskHandler func(handler *taskHandler)
```
hi @Sagi-levi,
please make sure to document all exported funcs ✌️

Sagi-levi replied:
Done successfully.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -0,0 +1,212 @@
+)
+
+// DefaultTaskChannelBuffer defines the default size of the task channel buffer used in TaskSchedule initialization.
+// 20 Was chosen to allow a minimal but reasonable buffer
```
hi @Sagi-levi,
please make sure all comments are written as complete sentences. 
for example:

> // 20 Was chosen to allow a minimal but reasonable buffer.

instead of:

> // 20 Was chosen to allow a minimal but reasonable buffer

Sagi-levi replied:
Fixed.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -0,0 +1,212 @@
+	resultsMutex          sync.Mutex
+	resultsWg             sync.WaitGroup
+	wg                    sync.WaitGroup
+	shutDownOnce          sync.Once
```
hi @Sagi-levi, 
consider naming `shutdown` instead of `shutDownOnce` to simplify var name for better readability.

Sagi-levi replied:
Done.

arieltod reviewed (changes requested):
Hi @Sagi-levi great job!

I left some comments, please read them carefully and if you have doubt regarding one of them talk to me. We already discussed some of the comments in our previous session but they are still relevant.

Some of these comments were posted for specific changed files but are relevant for the whole package. 

---

❗  Currently when pulling this PR and trying to run the code- it fails with error:

```
# scheduler/scheduler
scheduler/higherOrder.go:14:8: task.Name undefined (type *taskHandler has no field or method Name, but does have field name)

```

Please fix this before submitting to review again.

---

I'm attaching here some additional general comments to fix:

- [x]  Let's update the PR title to adhere to our guidelines - https://docs.onelayer.dev/pull-request.html#title
- [x]  Let's rephrase the commit titles o adhere to our guidelines (titles are not lower case, also multiple commits use the vague verbs of "adding", "implementing")- https://docs.onelayer.dev/pull-request.html#commits
- [x]  Fix the warnings thrown in IDE
- [x]  Review the package’s documentation via local `pkgsite`
- [ ]  Document all exported elements in the package
- [ ]  Justify all exported elements (and the opposite- don’t export unused elements in the package)
- [ ]  Reconsider naming of elements in the package

thanks 🧀

arieltod commented on main.go:
```diff
@@ -0,0 +1,29 @@
+package main
+
+import (
+	"log/slog"
```
These imports are throwing warning in IDE, need to sort them.
Please make sure to verify run `go fmt` on your committed files (either by configuring your IDE or by running the command line tool).
This comment is relevant for all files in this repo 🤞

Sagi-levi replied:
Fixed my IDE settings by matching them to yours, afterward fixed all imports as expected.

arieltod commented on go.mod:
```diff
@@ -0,0 +1,5 @@
+module task.scheduler
```
Hi @Sagi-levi ,
regarding the module naming- the dot in `task.scheduler` is typically reserved for domain-like prefixes in module paths (e.g., `go.onelayer.dev/x/component`), not within the descriptive component. 
Since our module is currently independent and not designed to be fetched as a dependency, a simpler path like `task-scheduler` is valid. 


```suggestion
module task-scheduler
```

https://go.dev/doc/modules/layout
https://go.dev/ref/mod

arieltod commented on scheduler/scheduler_test.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+func TestScheduler(t *testing.T) {
+	t.Run("test type", func(t *testing.T) {
```
How checking the type tests our scheduler's logic? 
I think these checks are irrelevant for the scheduler package and test more built-in go logic.

Sagi-levi replied:
Leftovers of deep TDD, removed!
Thanks for noticing.

arieltod commented on scheduler/task.go:
```diff
@@ -0,0 +1,16 @@
+type taskHandler struct {
+}
+
+func (t *taskHandler) run(ctx context.Context) error {
```
Why do we need context here if it's not used?

arieltod commented on scheduler/task.go:
```diff
@@ -0,0 +1,16 @@
+)
+
+// Task defines the interface to implement by an end user for a task that can be executed by the scheduler.
+type Task interface {
```
Why do we need empty funcs and interface here? 
Let's drop the skeleton naming, and in our initial commit of the scheduler package provide a simple task scheduler.

Sagi-levi replied:
New implementation demands a function that returns an error, no unnecessary structs and abuse.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -1,19 +1,100 @@
+	var errMsgs []string
+
+	if taskChannelBuffer <= 0 {
+		taskChannelBuffer = defaultTaskChannelBuffer
```
IMO the pattern of using const value when the inputed value from user is invalid is not the best,
I think it will be better to thrown an error in case the user inputed negative value here.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -1,19 +1,100 @@
+	}
+	if concurrentWorkerLimit <= 0 {
+		concurrentWorkerLimit = defaultConcurrentWorkerLimit
+		errMsgs = append(errMsgs, "concurrent worker limit must be greater than 0")
```
Check go's errors [package](https://pkg.go.dev/errors#Join), `errors.join` already wraps multiple error messages.

Sagi-levi replied:
Fixed functionality to: in case of missmatch between reasonable limit or buffer, an error will return.
Errors now handed with join

arieltod commented on scheduler/task.go:
```diff
@@ -2,15 +2,20 @@ package scheduler
 }
 
 type taskHandler struct {
+	id   uuid.UUID
```
why are we using `uuid` here and not `int`?
I assume it's because we want to have unique id values. If so- please explain this in a comment.

Sagi-levi replied:
Done.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -1,19 +1,100 @@
+	wg                    sync.WaitGroup
+	shutDown              sync.Once
+	startRunning          sync.Once
+	registered            atomic.Int32
```
@Sagi-levi WDYT about using the channel's `len` func to check how many tasks are registered? 
As we talked about- we're assuming that the user cannot register additional tasks if the scheduler is already started running. 

I think we can drop these counters. 
But in case you do keep them and there are multiple counters let's create a custom type like `runCounters` that will hold all this info so we'll be able to use it with:

```go
s := NewTaskSchedule(c,1,2)
s.logger("done", s.runCounters.done)
s.logger("failed", s.runCounters.failed)
```

Sagi-levi replied:
```go
	registered            int
	runCounters           runCounters
}
type runCounters struct {
	done   atomic.Int32
	failed atomic.Int32
}
```
Thanks for suggesting this improvement!
 To provide insights at summary, metrics collection in a thread safe way is a must.

arieltod commented on scheduler/higherOrder.go:
```diff
@@ -0,0 +1,28 @@
+	}
+}
+
+// WithName sets a custom name for a task by modifying its `taskHandler` configuration.
```
@Sagi-levi please verify to fill lines when writing long comments. 
Use `option`+`enter` keys in your mac. 
<img width="1057" height="201" alt="image" src="https://github.com/user-attachments/assets/1bce84d3-e2e0-4d0a-8d56-e301f5d4458d" />

❗ this comment is relevant for all commits & files in this PR

arieltod commented on scheduler/myTask.go:
```diff
@@ -0,0 +1,19 @@
+)
+
+// MyTask represents a task or unit of work that can be executed within a given context.
+type MyTask struct {
```
Why do we need `MyTask`?
It's used only in `main.go` and unitests. 
I think that the (missing) usages point that this type is redundant. Let's use the `Task` interface instead.

Sagi-levi replied:
Removed completely due to change of task interface.

arieltod commented on scheduler/scheduler_test.go:
```diff
@@ -0,0 +1,155 @@
+		}
+	})
+	t.Run("test type error", func(t *testing.T) {
+		scheduler1, _ := NewTaskSchedule(5, 40)
```
@Sagi-levi why are we ignoring the errors? 
It points that we're might not testing the real logic of the scheduler and specifically of `NewTaskSchedule`.

Sagi-levi replied:
Added an error handling to the tests where needed.

arieltod commented on scheduler/scheduler_test.go:
```diff
@@ -0,0 +1,155 @@
+
+// Example demonstrates the creation, registration, execution, and summary of tasks using the TaskSchedule infrastructure.
+// Dropping the time key provides us consistent output.
+func Example() {
```
Please make sure the give indicative name for the example tests.
```suggestion
func ExampleScheduler() {
```

Sagi-levi replied:
Fixed!

arieltod commented on scheduler/scheduler_test.go:
```diff
@@ -0,0 +1,155 @@
+	logHandler := slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
+		ReplaceAttr: func(_ []string, a slog.Attr) slog.Attr {
+			if a.Key == slog.TimeKey {
+				return slog.Attr{} // drop the time field
```
Why are dropping the time field? 
Please make sure to rephrase the code comment as a complete sentence (missing comma).

Sagi-levi replied:
Changed.

arieltod commented on go.mod:
```diff
@@ -0,0 +1,3 @@
+module task.scheduler
+
+go 1.24.4
```
hi @Sagi-levi,
please make sure that the commit's domain (the first part, before ":") is `go.mod` for these type of commits:

> go.mod: initialize task-scheduler module 

For example check this [reference](https://github.com/OneLayerHQ/vendor-multitech/commit/6f88675847a124f61fffced5521c35074e193bcb).

Also- there's no need in the commit's description let's drop it.

Sagi-levi replied:
Changed as suggested.

arieltod commented on .gitignore:
```diff
@@ -0,0 +1,20 @@
+# vendor/
+
+# Go workspace file
+go.work
```
Hi @Sagi-levi,
The commit's title should start with domain "gitignore", check this reference:
https://github.com/OneLayerHQ/goflow2/commit/c1dfc55d5ed62638734f2a59f9c1c1bfb4623f43

Also- please add informative commit description why are we ignoring these files.

Sagi-levi replied:
Merged with the init file and fixed the module name.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -57,9 +60,19 @@ func NewTaskSchedule(taskChannelBuffer, concurrentWorkerLimit int) (*TaskSchedul
 
 // Register adds a task or process to the TaskSchedule for execution as per its scheduling configuration.
-func (t *TaskSchedule) Register(task taskHandler) {
+func (t *TaskSchedule) Register(task Task, caps ...HighOrderTaskHandler) {
```
Consider renaming the high order handler to:
```suggestion
func (t *TaskSchedule) Register(task Task, opts ...Options)
```

the `...options` is a common pattern in go, check this:
https://medium.com/@omid.jn/the-option-pattern-in-go-9edf55e88320

Sagi-levi replied:
Fixed.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -74,12 +87,14 @@ func (t *TaskSchedule) Start() {
-						t.failedRun.Add(1)
+					for i := 0; i < task.retries; i++ {
+						t.currentRun.Add(1)
+						if err := task.run(context.Background()); err != nil {
```
Why are passing here context if not used later?
Let's simplify the code by removing it.

Sagi-levi replied:
Removed.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -54,6 +58,7 @@ func NewTaskSchedule(taskChannelBuffer, concurrentWorkerLimit int) (*TaskSchedul
 
 	return &TaskSchedule{
 		taskChan:              make(chan taskHandler, taskChannelBuffer),
+		resChan:               make(chan res, taskChannelBuffer),
```
What is the diff between task and done channel? the names are a bit confusing, in this case even though they are not exported we can still add a comment.

Sagi-levi replied:
Guessed you meant to rechannel, I split the commit of name to 2 commits and explained about the results collecting.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -82,25 +87,44 @@ func (t *TaskSchedule) Register(task Task, caps ...HighOrderTaskHandler) {
 						t.currentRun.Add(1)
 						if err := task.run(context.Background()); err != nil {
 							t.failedRun.Add(1)
+							result.isOK = false
```
FYI- since `isOK` is bool type, the zero value is `false` by default

Sagi-levi replied:
Fixed.

arieltod commented on scheduler/task.go:
```diff
@@ -26,3 +30,11 @@ type res struct {
 	isOK bool
 }
+
+func (t *taskHandler) log(format string, a ...any) {
```
why do we need this wrapper?
why can't we print logs with `t.logger.Info` outside this func?

Sagi-levi replied:
We ain't going to use logger outside the package.  Only the necessary logs are printed, the logic of with logging provides as a way to insert a logger as we like (can be examined on test file and main.go file). This way we control printing in terms of when and how.

arieltod commented on scheduler/scheduler.go:
```diff
@@ -1,19 +1,100 @@
+	taskChan              chan taskHandler
+	doneChan              chan struct{}
+	wg                    sync.WaitGroup
+	shutDown              sync.Once
```
Why do we need the `sync.Once`?
Don't forget that as the writer of this package you have the power to tell the user how to use the package 👼 .

Sagi-levi replied:
Removed, added an explanation in docs file.

arieltod reviewed (changes requested):
hi @Sagi-levi,

I left some more comments in the PR. 
In addition there are few comments from last time that are still relevant that we discussed about (PR title, commit titles, atomic commits, package + module naming etc.). 
Please address them and let me know if you any questions.

Thanks! 🧀

Sagi-levi committed:
taskscheduling: construct a new instance of task scheduler

This commit creates a new scheduler according to the user's needs, to promise
full control and safe usage of the package. In case of invalid inputs, the user
will get an error.

Sagi-levi committed:
go.mod: go get github.com/google/uuid

Sagi-levi committed:
taskscheduling: log task behavior

This commit includes the ability to log tasks behavior around all over the task
life cycle, providing the user with simple insights about his task.

Sagi-levi committed:
taskscheduling: present a human-readable summary report

This commit creates a summary report to the user, after closing the scheduler,
showing the user what happened during the task running.

Sagi-levi committed:
taskscheduling: illustrate an example of taskscheduling

Example of the scheduler functionality, specifically here we create a logger
that doesn't write time, to make the result consistent for the output.

Sagi-levi committed:
taskscheduling: showcase concurrent tasks scheduler

This commit is showing the common user the optional usage of the task scheduler
package. Creating a new scheduler, creating tasks, registering the tasks to the
scheduler and run, later stoping it and showing the results.

Sagi-levi committed:
taskscheduling: register new tasks to the scheduler

We are providing the user a function to register his tasks to the scheduler.
In addition, We're implementing the taskHandler to capture metadata about tasks
and their behavior.

Sagi-levi committed:
taskscheduling: run the scheduler

This commit allows the user to run the tasks he registered to the scheduler.
Because we are running concurrently, we use atomic counters, channels, and go
routines.

Sagi-levi committed:
taskscheduling: stop the scheduler

In this commit we stop the scheduler running, this is to allow full control over
the scheduler.

Sagi-levi committed:
taskscheduling: result information of task scheduler

This commit collects the result of the running task during the execution, using
thread safe abilities.

Sagi-levi committed:
taskscheduling: name tasks

This commit includes the ability to name tasks as the user wants, giving the
user more control over his tasks.

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -0,0 +1,38 @@
+	concurrentWorkerLimit int
+}
+
+// NewScheduler creates and returns a Scheduler with specified taskChannelBuffer
```
WDYT about this rephrasing?

```suggestion
// NewScheduler creates and returns a Scheduler with specified configurations of the size of tasks channel
// and the concurrent worker limit. For invalid values an error will be returned.
```

Sagi-levi replied:
WDYT about 
```
// NewScheduler creates and returns a Scheduler with the specified task channel size...
``` 
?
I prefer to avoid 2x "of".
@arieltod

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -36,3 +39,19 @@ func NewScheduler(taskChannelBuffer, concurrentWorkerLimit int) (*Scheduler, err
 		concurrentWorkerLimit: concurrentWorkerLimit,
 	}, nil
 }
+
```
hi! check the typo in the commit message description:

> we create implement the taskHandler so we can have metadata about
> tasks and their behavior.

Sagi-levi replied:
How this sounds:
```
we implement the taskHandler to capture metadata about
tasks and their behavior.
``` 
?
@arieltod

arieltod replied:
> We're implementing the taskHandler to capture metadata about
> tasks and their behavior.

Sagi-levi replied:
Will do.

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -15,6 +16,12 @@ type Scheduler struct {
 	doneChan              chan struct{}
 	wg                    sync.WaitGroup
 	concurrentWorkerLimit int
+	registered            int
```
If `registered` is a counter too, why not define it as `runCounters` member? 👼

Sagi-levi replied:
The reason that registered counter is outside the runCounters is because of 2 reasons.
1. It doesn't relate to the "running", only the registration.
2. It is a regular INT, not atomic one. Since we limited the user usage of the package, we don't need to keep the registration INT thread-safe.

In case you still think it's better to join them, I will.
@arieltod

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -86,3 +86,9 @@ func (s *Scheduler) stopRunSafe() {
 }
+
+// Stop the execution of the Scheduler, stopping any ongoing tasks or processes.
+func (s *Scheduler) Stop() {
```
hi @Sagi-levi, I was expecting to see in this commit a unitest that uses this func too 🤞

Sagi-levi replied:
Done!

arieltod commented on taskscheduler/doc.go:
```diff
@@ -2,6 +2,12 @@
 
 This package enables you to schedule and execute tasks concurrently.
+
+Guidelines of this package are that you can call the Run Function only once and
```
```suggestion
The guidelines of this package include:
- Each scheduler can be run only once.
- Tasks must be registered before their scheduler starts running.
- Stop() can be called only after Run() was called.
```

The better practice when writing doc of packages is not avoid using bullet points but in this case it improves the readability. 
For example the sim management doc [file](https://github.com/OneLayerHQ/atmosphere/blob/b82cc1b3d72bbe8344ca870e3f54bef4b2297ae5/simmgmt/doc.go).

Sagi-levi replied:
Fixed.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -37,3 +37,27 @@ func TestWithRetry(t *testing.T) {
+		s.Stop()
+		results := s.getResults()
+		if len(results) == 0 {
+			t.Errorf("got %v; expected not %v", results, results)
```
`results` is printed twice here 🤔.
Consider printing in this error the expected results length.

Sagi-levi replied:
Fixed.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,167 @@
+)
+
+func TestScheduler(t *testing.T) {
+	s, err := NewScheduler(5, 10)
```
what about testing invalid config values?

Sagi-levi replied:
Done.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,167 @@
+func TestScheduler(t *testing.T) {
+	s, err := NewScheduler(5, 10)
+	if err != nil {
+		t.Error(err)
```
```suggestion
		t.Fatalf("%v", err)
```

Let's use `t.Fatalf` when we're detecting an issue that should immediately fail and terminate the test.

Sagi-levi replied:
Done.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,167 @@
+func TestWithLogging(t *testing.T) {
+	tests := []struct {
+		name              string
+		shouldFindMessage string
```
```suggestion
		expectedMsg string
```

Sagi-levi replied:
Done!

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -0,0 +1,205 @@
+	results               []res
+	resultsMutex          sync.Mutex
+	resultsWg             sync.WaitGroup
+	wg                    sync.WaitGroup
```
I think we can think of better naming for this member :) 
Make sure to name it so it will be clear what is the diff between `resultsWg` and `wg`.

Sagi-levi replied:
Changed to tasksWg.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,167 @@
+		fmt.Println("test")
+		return nil
+	}
+	s.Register(fn)
```
what about testing the num of registered tasks here?

Sagi-levi replied:
Done.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,167 @@
+)
+
+func TestScheduler(t *testing.T) {
+	s, err := NewScheduler(5, 10)
```
What happens if you register more tasks than the worker limit? 
It should be tested too.

Sagi-levi replied:
Done.

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -0,0 +1,205 @@
+	}
+	go s.stopRunSafe()
+}
+func (s *Scheduler) taskRepeater(task taskHandler) {
```
`taskRepeater` is relevant for all tasks? Including the once that are without retries? 
If so- the naming of the func is a bit confusing.

Sagi-levi replied:
According to line 64 of this file, every registration function call sets retries to 1 by default. Adding code that checks if the default value hasn't changed and sets it in the running function would unnecessarily complicate the code.

However, this made me realize that I never validate what the user sets when using the retry option. I'll add validation for this now!
@arieltod

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+func TestWithLogging(t *testing.T) {
+	tests := []struct {
```
@Sagi-levi I don't see any logs when running this test:
```shell
=== RUN   TestWithLogging
=== RUN   TestWithLogging/captures_messages_exist

--- PASS: TestWithLogging/captures_messages_exist (0.00s)
=== RUN   TestWithLogging/captures_messages_not_exist

--- PASS: TestWithLogging/captures_messages_not_exist (0.00s)
--- PASS: TestWithLogging (0.00s)
```

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -0,0 +1,205 @@
+	go s.stopRunSafe()
+}
+func (s *Scheduler) taskRepeater(task taskHandler) {
+	for i := 0; i < task.retries; i++ {
```
So we are retrying tasks even if they succeeded?
We need to retry only if it fails. 

From the README [file](https://github.com/OneLayerHQ/onboarding-exercises/tree/main/exercises/schedulertask#readme):
> Implement a higher-order function, `WithRetry`, that takes a task and a retry count as arguments and returns a new task that retries the original task a specified number of times if it fails.

Sagi-levi replied:
Fixed and tested.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -12,3 +13,37 @@ func TestSchedulerValidation(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// Fill the channel completely
```
```
// Fill the channel completely.
```

Sagi-levi replied:
Done.

arieltod commented on taskscheduler/scheduler.go:
```diff
@@ -15,6 +16,12 @@ type Scheduler struct {
+	registered            int
+	runCounters           runCounters
+}
+type runCounters struct {
```
missing line here ✋ 

also, if `runCounters` are added in this commit to count done & failed it should also affect the unitests of the package.

Sagi-levi replied:
Without a proper stop the result won't be atomic, I added a test after that checks the runCounters fail and done in the stop commit.

arieltod reviewed (changes requested):
hi @Sagi-levi 👋 
I left some comments in the meantime but let's talk more about the PR tomorrow in the office!

arieltod commented on taskscheduler/task.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/google/uuid"
+)
+
+type Task func() error
```
Please add doc to Exported funcs 🤞

Sagi-levi replied:
Done.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,249 @@
+	}
+}
+func TestWithName(t *testing.T) {
+	t.Run("test with name", func(t *testing.T) {
```
Why do we need subtest here?

Sagi-levi replied:
Done.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,249 @@
+		return nil
+	}
+	s.Register(fn1, WithRetry(retry))
+	s.Register(fn2)
```
`fn1` & `fn2` are dups,
this will work here too:

```
s.Register(fn1, WithRetry(retry))
s.Register(fn1)
```

Sagi-levi replied:
For some reason, my fn1 should return err but didn't, now it's fixed.

arieltod commented on taskscheduler/scheduler_test.go:
```diff
@@ -0,0 +1,249 @@
+	s.Register(fn2)
+	s.Run()
+	s.Stop()
+	if s.runCounters.done.Load() != int32(retry)+1 {
```
why +1? 
Pls add comment to explain.

Sagi-levi replied:
Added a comment.

arieltod commented on taskscheduling/scheduler.go:
```diff
@@ -0,0 +1,223 @@
+	registered            int
+	runCounters           runCounters
+	stopped               atomic.Bool
+}
```
@Sagi-levi  missing new line here

Sagi-levi replied:
Done.

arieltod commented on main.go:
```diff
@@ -0,0 +1,49 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	task1 := func() error {
```
Why do you need to initialize these duplicated tasks? `task1`, `task2` and `task3`
You can register multiple tasks that will run the same func.

Sagi-levi replied:
The goal is to mimic the end user of the package. I chose to create different tasks but without inner implementation. "Simplifying" the code even more, in my opinion, could mislead the user.

arieltod replied:
IMO it's not a good example for the package usage. 
Let's remove the duplicate vars here.

arieltod commented on taskscheduling/scheduler.go:
```diff
@@ -0,0 +1,223 @@
+// Scheduler manages the scheduling, execution, and reporting of tasks with
+// controlled concurrency.
+type Scheduler struct {
+	taskChan              chan taskHandler
```
Hi @Sagi-levi,
Please refrain from using the type in the member name 🤞

Sagi-levi replied:
HI @arieltod,
From a quick look over the atmosphere repo of one layer, I noticed that there are several usages of naming channels with the end "chan", if still that's the standard I will align with it.

https://github.com/OneLayerHQ/atmosphere/blame/7165296c32cd9a66ac02fb97a6761b3f23b99676/simmgmt/activity/controlregistry.go#L19
https://github.com/OneLayerHQ/atmosphere/blame/7165296c32cd9a66ac02fb97a6761b3f23b99676/visibility/cpepolling/instrument.go#L542
https://github.com/OneLayerHQ/atmosphere/blame/7165296c32cd9a66ac02fb97a6761b3f23b99676/visibility/cellular/relayfilter/internal/atomicset/atomicset.go#L114

arieltod replied:
Yep all these references are very old code that was migrated into this repo by @ofektavor. 
We're trying our best to keep up the high coding standarts in our repos but it's not always possible. 

IMO go's built package have better references:

https://cs.opensource.google/go/go/+/master:test/ken/chan.go;drc=eb37b5b74499c1c5f90a1adf533dc59fa870d794;l=25
https://cs.opensource.google/go/go/+/master:test/chan/powser2.go;drc=93bcf912994931780cdd4f5492ae8fcfdf760208;l=47

arieltod commented on taskscheduling/scheduler.go:
```diff
@@ -0,0 +1,223 @@
+	concurrentWorkerLimit int
+	registered            int
+	runCounters           runCounters
+	stopped               atomic.Bool
```
In Go, a common convention is to use positive and descriptive names that often start with `is`, `has`, `can`, etc., for boolean variables.

```suggestion
	inProgress               atomic.Bool
	isActive                    atomic.Bool
	isRunning                 atomic.Bool
```

Sagi-levi replied:
Done.

arieltod commented on taskscheduling/scheduler.go:
```diff
@@ -0,0 +1,223 @@
+}
+type result struct {
+	buffer  chan res
+	results []res
```
hi @Sagi-levi,
I think we can come up with better naming. 
The "res" pattern is too repetitive here and confusing. 

```suggestion
		s.result.all = append(s.result.all, res)
```

```suggestion
		s.result.arr = append(s.result.arr, res)
```

Sagi-levi replied:
What about 
result->resultHandler
res->result
results->all

arr remindes me of using the type as name and we said we want to avoid it.
@arieltod

arieltod commented on taskscheduling/scheduler.go:
```diff
@@ -0,0 +1,228 @@
+		},
+		concurrentWorkerLimit: concurrentWorkerLimit,
+	}
+	scheduler.InProgress.Store(true)
```
Are we really in progress if we initialised a new scheduler? 
If so- the naming is not accurate enough and you need to explain this either in comment or doc of the package.

Sagi-levi replied:
Default value of atomic bool is false, inProgress flag added to distinguish between stopped scheduler and non-stopped(FYI the flag name was stopped, but changed to meet go positive bool standards), to ensure registration is welcome we must set this flag to true, and it also fits our agenda of the scheduler isn't stopped, so he is in progress.
The channels aren't closed yet.

arieltod replied:
Also- why is it exported?

arieltod replied:
Renaming the var to "positive" naming in a good practice in go, but I think we can delete this row from `New()` since:
> Tasks must be registered before their scheduler starts running.

Also in 
```go
// Register adds a task or process to the Scheduler for execution as per its
// scheduling configuration.
func (s *Scheduler) Register(task Task, opts ...Opts) error {
	if !s.InProgress.Load() {
		return errors.New("scheduler is stopped")
	}
```

we already assume that when task is registered the scheduler didn't start running.

Sagi-levi replied:
Based on our meetings and the exercise, we discussed that tasks cannot be registered after Run execution. While I eventually agreed with removing support for registration after running, I believe we shouldn't let users encounter fatal errors due to misunderstanding how to use the package correctly. My common sense tells me that allowing users to have issues with their code isn't a good approach.

Changed form InProgress to inProgress. @arieltod

arieltod replied:
IMO it's not mandatory because of the doc.go guidelines for the package.
You need to decide what is the balance you want when enabling the user such actions & how to handle them. 
I'm leaving it up to you to decide how to handle these scenarios.

Sagi-levi replied:
After careful consideration, I decided that preventing user panic is the right approach. Users don't necessarily know whether I'm using channels or not (I hope they read the documentation but won't examine the code implementation).

Based on this assessment, I want to inform users about errors, prevent them from misusing the system, and let them decide how to respond. I will enforce this on register but not on run and stop, because run and stop are core functions that have more influence on the system.

arieltod replied:
Ack.
Don't forget to add a comment about this- that the panic will be thrown only in case of the user's package misuse!

Sagi-levi replied:
Fin.?
Done.
Added at docs and at the start and stop functions.

arieltod commented on taskscheduling/scheduler.go:
```diff
@@ -0,0 +1,228 @@
+	<-s.doneChan
+}
+
+func (s *Scheduler) stopRunSafe() {
```
What is the diff between `Stop()` and `stopRunSafe()`?
Sounds like `Stop()` is not safe 👀 

@Sagi-levi let's merge these funcs into a single `Stop()` func.

Sagi-levi replied:
Merging them won't serve the cause that they are made for.
stopRunSafe is for waiting to tasks wait group and results wait group, it needs to run at the end of run function.

Stop function is for enforcing progress stopping and closing channels.

I can change stopRunSafe to runCleanUp.
@arieltod What do you think?

arieltod replied:
When a reviewer sees 2 funcs that stop the scheduler, but only one of them is "safe" it's confusing.

Since `stopRunSafe` is used only in the go routine call in `Run()` let's put this code there and delete `stopRunSafe()`. 
Or rename it to more differentiated func name like `CleanUp()` or `Clear()`.

arieltod commented on taskscheduling/scheduler.go:
```diff
@@ -0,0 +1,228 @@
+func (s *Scheduler) Run() {
+	s.result.wg.Add(1)
+	go s.resultHandler()
+	s.registered = len(s.taskChan)
```
Why is it safe to assign `len(s.taskChan)` to `s.registered` here?
Let's add some comment to explain this.

arieltod commented on taskscheduling/scheduler_test.go:
```diff
@@ -0,0 +1,209 @@
+	"strings"
+	"testing"
+)
+
```
@Sagi-levi this file is a bit hard to read even though it uses subtests and test cases, I think we can improve the tests. 

IMO we should have 3 test funcs:
1. `TestConfig`- that use New() and validated the behaviour according to different configs.
2. `TestStop`- that stops the scheduler and verifies the expected behaviour
3. `TestScheduler` that runs the scheduler 

<details><summary>TestConfig</summary>
<p>

```go
func TestConfig(t *testing.T) {
	tests := []struct {
		name                  string
		taskChannelBuffer     int
		concurrentWorkerLimit int
		tasksToRegister       int
		expectedError         string
		shouldCreateScheduler bool
	}{
		{
			name:                  "negative buffer and worker limit",
			taskChannelBuffer:     -4,
			concurrentWorkerLimit: -6,
			expectedError:         "task channel buffer size must be greater than 0",
			shouldCreateScheduler: false,
		},
		{
			name:                  "zero buffer size",
			taskChannelBuffer:     0,
			concurrentWorkerLimit: 5,
			expectedError:         "task channel buffer size must be greater than 0",
			shouldCreateScheduler: false,
		},
		{
			name:                  "zero worker limit",
			taskChannelBuffer:     5,
			concurrentWorkerLimit: 0,
			expectedError:         "concurrent worker limit must be greater than 0",
			shouldCreateScheduler: false,
		},
		{
			name:                  "valid config but too many tasks",
			taskChannelBuffer:     2,
			concurrentWorkerLimit: 1,
			tasksToRegister:       3,
			expectedError:         "channel is full",
			shouldCreateScheduler: true,
		},
		{
			name:                  "large buffer with excessive tasks",
			taskChannelBuffer:     5,
			concurrentWorkerLimit: 2,
			tasksToRegister:       8,
			expectedError:         "channel is full",
			shouldCreateScheduler: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			s, err := New(tt.taskChannelBuffer, tt.concurrentWorkerLimit)

			if !tt.shouldCreateScheduler {
				if err == nil {
					t.Fatal("expected error, got none")
				}
				if !strings.Contains(err.Error(), tt.expectedError) {
					t.Errorf("expected error containing '%s', got '%s'", tt.expectedError, err.Error())
				}
			} else {
				if err != nil {
					t.Fatalf("unexpected error creating scheduler: %v", err)
				}

				for i := 0; i < tt.taskChannelBuffer; i++ {
					err = s.Register(func() error { return nil })
					if err != nil {
						t.Fatalf("unexpected error registering task %d: %v", i, err)
					}
				}

				err = s.Register(func() error { return nil })
				if err == nil {
					t.Fatal("expected error when buffer is full, got none")
				}
				if !strings.Contains(err.Error(), tt.expectedError) {
					t.Errorf("expected error containing '%s', got '%s'", tt.expectedError, err.Error())
				}
			}
		})
	}
}
``` 

</p>
</details> 

<details><summary>TestStop</summary>
<p>

```go
func TestStop(t *testing.T) {
	s, err := New(5, 2)
	if err != nil {
		t.Fatal(err)
	}

	var completedTasks atomic.Int32
	taskDuration := 200 * time.Millisecond

	// Register slow tasks
	for i := 0; i < 4; i++ {
		err = s.Register(func() error {
			time.Sleep(taskDuration)
			completedTasks.Add(1)
			return nil
		})
		if err != nil {
			t.Fatal(err)
		}
	}

	s.Run()

	// Wait a bit to let some tasks start, then stop
	time.Sleep(100 * time.Millisecond)
	s.Stop()

	// Verify some tasks completed but not all
	completed := completedTasks.Load()
	done := s.runCounters.done.Load()

	if completed == 0 {
		t.Error("expected at least some tasks to complete")
	}

	if completed != done {
		t.Errorf("completed counter mismatch: atomic=%d, scheduler=%d", completed, done)
	}

	if done > 4 {
		t.Errorf("expected max 4 done tasks, got %d", done)
	}
}
```

</p>
</details> 

<details><summary>TestScheduler</summary>
<p>

```go
func TestScheduler(t *testing.T) {
	s, err := New(5, 2)
	if err != nil {
		t.Fatal(err)
	}

	testLogger, buf := createTestLogger()
	var attemptCount atomic.Int32
	successOnAttempt := int32(3)

	// Register successful task with name
	taskName := "success-task"
	err = s.Register(func() error { return nil }, WithName(taskName))
	if err != nil {
		t.Fatal(err)
	}

	// Register task with retry logic that succeeds on nth try, with logging
	err = s.Register(func() error {
		count := attemptCount.Add(1)
		if count < successOnAttempt {
			return errors.New("fail on purpose")
		}
		return nil
	}, WithRetry(5), WithLogging(testLogger), WithName("retry-task"))
	if err != nil {
		t.Fatal(err)
	}

	// Register task that always fails with retries
	err = s.Register(func() error {
		return errors.New("always fails")
	}, WithRetry(2), WithName("fail-task"))
	if err != nil {
		t.Fatal(err)
	}

	// Register another successful task (no name)
	err = s.Register(func() error { return nil })
	if err != nil {
		t.Fatal(err)
	}

	s.Run()
	s.Stop()

	// Verify registered count
	if s.registered != 4 {
		t.Errorf("expected 4 registered tasks, got %d", s.registered)
	}

	// Verify task execution counts
	// retry-task: 2 failures + 1 success = 3 executions
	// fail-task: 2 failures = 2 executions
	// success-task: 1 success = 1 execution
	// unnamed task: 1 success = 1 execution
	// Total: 7 executions, 4 failures
	expectedDone := int32(7)
	expectedFailed := int32(4)

	if s.runCounters.done.Load() != expectedDone {
		t.Errorf("expected %d done tasks, got %d", expectedDone, s.runCounters.done.Load())
	}

	if s.runCounters.failed.Load() != expectedFailed {
		t.Errorf("expected %d failed tasks, got %d", expectedFailed, s.runCounters.failed.Load())
	}

	if attemptCount.Load() != successOnAttempt {
		t.Errorf("expected %d attempts for retry task, got %d", successOnAttempt, attemptCount.Load())
	}

	// Verify successful task count
	successCount := 0
	taskNameFound := false
	for _, result := range s.result.results {
		if result.isOk {
			successCount++
		}
		if result.task.name == taskName && result.isOk {
			taskNameFound = true
		}
	}

	if successCount != 3 {
		t.Errorf("expected 3 successful tasks, got %d", successCount)
	}

	if !taskNameFound {
		t.Errorf("task with name '%s' not found or not successful", taskName)
	}

	// Verify logging captured messages
	output := buf.String()
	expectedMessages := []string{"registered", "running task", "retry-task"}
	for _, msg := range expectedMessages {
		if !strings.Contains(output, msg) {
			t.Errorf("expected log to contain '%s', got: %s", msg, output)
		}
	}

	// Verify we can't register after stop
	err = s.Register(func() error { return nil })
	if err == nil {
		t.Fatal("expected error when registering after Stop()")
	}
	if !strings.Contains(err.Error(), "scheduler is stopped") {
		t.Errorf("expected 'scheduler is stopped' error, got: %v", err)
	}
}
```

</p>
</details>

Sagi-levi replied:
TestConfig does more than just testing configuration - it also runs and register. TestStop performs similar operations that overlap with what TestConfig already does, creating duplication. Our goal is to make the code more efficient by removing duplicates and making it shorter.
Also there are no logging beside errors.

Let's review the expectations for this file. We can look at good examples from our codebase and create standards file in Notion based on them.

Sagi-levi replied:
New changes pushed! @arieltod

arieltod reviewed (changes requested):
hi @Sagi-levi,

I feel like we improved a lot in this PR in our latest review iterations 🥳 great job!

I left some more small comments, LMK if you want to discuss this further :)

arieltod commented on taskscheduling/scheduler_test.go:
```diff
@@ -0,0 +1,209 @@
+		}
+	})
+	t.Run("testing scheduler flow", func(t *testing.T) {
+		//Registering tasks to scheduler buffer to ensure registration is limited and
```
@Sagi-levi some comments in this file are not formatted according to guidelines
```suggestion
//Registering tasks to scheduler buffer to ensure registration is limited and
```

Sagi-levi replied:
Changed all around the code.
// Registering
instead of
//Registering

arieltod commented on taskscheduling/scheduler_test.go:
```diff
@@ -0,0 +1,209 @@
+		} else {
+			t.Log(`task name updated as expected`)
+		}
+		tests := []struct {
```
Unit test cases should be initialized in the top rows of the test func

Sagi-levi replied:
New changes pushed! @arieltod

Sagi-levi committed:
taskscheduling: retry tasks using "opts" pattern

This commit includes the ability to retry tasks as the user wants, giving the
user more control over his tasks.

arieltod commented:
Nice work done here!
You've made a great progress in writing this PR 🥳 

![gif](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExZTJ6MjFjN3pwcHgwOXlvbHhuMjFmbzNvbzZlYWFyOWNvdnNtaWo1ZyZlcD12MV9naWZzX3NlYXJjaCZjdD1n/bnzebvRguQVCNxF8KV/giphy.gif)

We can close this PR and move to the next one 🤞

Sagi-levi commented:
![alt text](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExMGZoaW9la29jMmpndmNyb2Fncm1kdWVpMmV0MDQ1c3RuMzV1bjJoaCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3ohfFyCp0wqtseckyQ/giphy.gif)
